File: app/path_display.py
import os

from app.media_roots import load_media_root_globals


PICTURES_ROOTS, PICTURES_ROOT_LABELS, PICTURES_ROOT_REAL_PATHS = load_media_root_globals()


def _label_for_root(root: str, index: int) -> str:
    if index < len(PICTURES_ROOT_LABELS):
        return PICTURES_ROOT_LABELS[index]
    return os.path.basename(root) or root.strip("/") or root


def _match_root(path: str):
    normalized = path.replace("\\", "/")
    roots = sorted(enumerate(PICTURES_ROOTS), key=lambda item: len(item[1]), reverse=True)
    for index, root in roots:
        root_normalized = root.replace("\\", "/").rstrip("/")
        if normalized == root_normalized or normalized.startswith(root_normalized + "/"):
            rel = normalized[len(root_normalized):].lstrip("/")
            return index, root_normalized, rel
    return None


def display_path(path: str) -> str:
    match = _match_root(path)
    if match:
        index, root_normalized, rel = match
        label = _label_for_root(root_normalized, index)
        return f"{label}/{rel}" if rel else label
    return path


def real_path(path: str) -> str:
    match = _match_root(path)
    if not match:
        return path

    index, _root_normalized, rel = match
    if index >= len(PICTURES_ROOT_REAL_PATHS):
        return path

    real_root = PICTURES_ROOT_REAL_PATHS[index].rstrip("/\\")
    if not real_root:
        return path
    if not rel:
        return real_root

    sep = "\\" if "\\" in real_root else "/"
    return real_root + sep + rel.replace("/", sep).replace("\\", sep)

File: rust/gallery_accel/src/path_display.rs
use std::path::Path;

use crate::media_roots::MediaRoots;

pub(crate) fn default_label(root: &str) -> String {
    let cleaned = root.trim_end_matches(['/', '\\']);
    Path::new(cleaned)
        .file_name()
        .and_then(|name| name.to_str())
        .map(|name| name.to_string())
        .filter(|name| !name.is_empty())
        .unwrap_or_else(|| cleaned.trim_matches(['/', '\\']).to_string())
}

pub(crate) fn display_path(path: &str, roots: &MediaRoots) -> String {
    let normalized = path.replace('\\', "/");
    // Match longest among virtual roots and real authorized roots.
    let mut candidates: Vec<(usize, String)> = Vec::new();
    for (index, root) in roots.roots.iter().enumerate() {
        let root_n = root.replace('\\', "/").trim_end_matches('/').to_string();
        if !root_n.is_empty() {
            candidates.push((index, root_n));
        }
        if let Some(real) = roots.real_root_at(index) {
            let real_n = real.replace('\\', "/").trim_end_matches('/').to_string();
            if !real_n.is_empty() {
                candidates.push((index, real_n));
            }
        }
    }
    candidates.sort_by(|a, b| b.1.len().cmp(&a.1.len()));
    for (index, root) in candidates {
        if normalized == root || normalized.starts_with(&(root.clone() + "/")) {
            let rel = normalized[root.len()..].trim_start_matches('/');
            let label = roots
                .labels
                .get(index)
                .cloned()
                .unwrap_or_else(|| default_label(&root));
            if rel.is_empty() {
                return label;
            }
            return format!("{}/{}", label, rel);
        }
    }
    path.to_string()
}
